Mathlib/Computability/DFA.lean

```diff
@@ -163,7 +172,80 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [Set.mem_singleton_iff] at ha' hc'
   substs ha' hc'
   have h := M.evalFrom_of_pow hb hb'
-  rwa [mem_accepts, evalFrom_of_append, evalFrom_of_append, h, hc]
+  rwa [mem_accepts, eval, evalFrom_of_append, evalFrom_of_append, h, hc]
 #align DFA.pumping_lemma DFA.pumping_lemma
 
+section Maps
+
+variable {α' σ' : Type*} (f : α' → α) (g : σ ≃ σ')
+
+/-- Transforms the alphabet of a DFA. -/
+@[simps]
+def comap : DFA α' σ where
+  step s a := M.step s (f a)
+  start := M.start
+  accept := M.accept
+
+@[simp]
+theorem comap_id : M.comap id = M := rfl
+
+@[simp]
+theorem evalFrom_comap (s : σ) (x : List α') :
+    (M.comap f).evalFrom s x = M.evalFrom s (x.map f) := by
+  induction x using List.list_reverse_induction with
+  | base => simp
+  | ind x a ih => simp [ih]
+
+@[simp]
+theorem eval_comap (x : List α') : (M.comap f).eval x = M.eval (x.map f) := by simp [eval]
+
+@[simp]
+theorem accepts_comap : (M.comap f).accepts = List.map f ⁻¹' M.accepts := by
+  ext x
+  conv =>
+    rhs
+    rw [Set.mem_preimage, mem_accepts]
+  simp [mem_accepts]
+
+/-- Lifts an equivalence on states to an equivalence on DFAs. -/
+@[simps apply_step apply_start apply_accept]
+def reindex : DFA α σ ≃ DFA α σ' where
+  toFun M := {
+    step := fun s a => g (M.step (g.symm s) a)
+    start := g M.start
+    accept := g.symm ⁻¹' M.accept
+  }
+  invFun M := {
+    step := fun s a => g.symm (M.step (g s) a)
+    start := g.symm M.start
+    accept := g ⁻¹' M.accept
+  }
```

Comment:
@digama0, this is a good example of situation where we really want nested `where` notation

Reply:
Or  `step s a := g.symm (M.step (g s) a)` notation within `{}`

Reply:
Yep, I remember trying both of those syntaxes and realizing neither work 😢